File: lua/grains/grains.lua
local grains = {}

local function subtotal(n, prev_grains)
    if n == 1 then
        return prev_grains;
    else
        return subtotal(n - 1, prev_grains + grains.square(n))
    end
end

function grains.square(n)
    return 2 ^ (n - 1)
end

function grains.total()
    return subtotal(64, 0)
end

return grains

File: go/grains/grains.go
package grains

import (
	"errors"
)

// Square returns the number of grains on the specified square of the chessboard.
// The number must be between 1 and 64 (inclusive). If the number is out of range,
// an error is returned.
func Square(number int) (uint64, error) {
	if number < minSquare || number > maxSquare {
		return 0, errors.New("number must be between 1 and 64, inclusive")
	}
	return pow2(number - 1), nil
}

// Total calculates the total number of grains on the chessboard,
// from the first square to the last one (square 64).
func Total() uint64 {
	return pow2(maxSquare) - 1
}

// pow2 calculates 2 raised to the power of the given number.
func pow2(number int) uint64 {
	return uint64(1 << number)
}

const minSquare, maxSquare = 1, 64
